dace/memlet.py

```diff
@@ -140,75 +248,178 @@ def simple(data,
+        self._state = state
+        self._edge = edge
 
-        return Memlet(data,
-                      na,
-                      subset,
-                      veclen,
-                      wcr=wcr,
-                      other_subset=other_subset,
-                      wcr_conflict=wcr_conflict,
-                      debuginfo=debuginfo)
+        # Find source/destination of memlet
+        try:
+            path = state.memlet_path(edge)
+        except (ValueError, AssertionError, StopIteration):
+            # Cannot initialize yet
+            return
+
+        is_data_src = True
+        if isinstance(path[-1].dst, AccessNode):
+            if path[-1].dst.data == self._data:
+                is_data_src = False
+        self._is_data_src = is_data_src
 
     @staticmethod
     def from_array(dataname, datadesc, wcr=None):
         """ Constructs a Memlet that transfers an entire array's contents.
             :param dataname: The name of the data descriptor in the SDFG.
             :param datadesc: The data descriptor object.
             :param wcr: The conflict resolution lambda.
-            @type datadesc: Data.
+            :type datadesc: Data
         """
-        range = subsets.Range.from_array(datadesc)
-        return Memlet(dataname, range.num_elements(), range, 1, wcr=wcr)
+        rng = subsets.Range.from_array(datadesc)
+        return Memlet.simple(dataname, rng, wcr_str=wcr)
 
     def __hash__(self):
-        return hash((self.data, self.num_accesses, self.subset, self.veclen,
-                     str(self.wcr), self.other_subset))
+        return hash(
+            (self.volume, self.src_subset, self.dst_subset, str(self.wcr)))
 
     def __eq__(self, other):
         return all([
-            self.data == other.data, self.num_accesses == other.num_accesses,
-            self.subset == other.subset, self.veclen == other.veclen,
-            self.wcr == other.wcr, self.other_subset == other.other_subset
+            self.volume == other.volume, self.src_subset == other.src_subset,
+            self.dst_subset == other.dst_subset, self.wcr == other.wcr
         ])
 
     def num_elements(self):
         """ Returns the number of elements in the Memlet subset. """
-        return self.subset.num_elements()
+        if self.subset:
+            return self.subset.num_elements()
+        elif self.other_subset:
+            return self.other_subset.num_elements()
+        return 0
 
     def bounding_box_size(self):
         """ Returns a per-dimension upper bound on the maximum number of
             elements in each dimension.
 
             This bound will be tight in the case of Range.
         """
-        return self.subset.bounding_box_size()
+        if self.src_subset:
+            return self.src_subset.bounding_box_size()
+        elif self.dst_subset:
+            return self.dst_subset.bounding_box_size()
+        return []
+
+    # New fields
+    @property
+    def src_subset(self):
```

Comment:
Ohhh I see what you're doing there :P sneaking in the new definition

Reply:
yep :) slowly but surely